blivet/zfcp.py

```diff
@@ -46,63 +47,117 @@ def logged_write_line_to_file(fn, value):
     def _to_string(self):
-        return "%s %s %s" % (self.devnum, self.wwpn, self.fcplun)
+        return str(self.devnum)
 
     def __str__(self):
         return stringize(self._to_string())
 
     def __unicode__(self):
         return unicodeize(self._to_string())
 
-    def online_device(self):
-        online = "%s/%s/online" % (zfcpsysfs, self.devnum)
-        portadd = "%s/%s/port_add" % (zfcpsysfs, self.devnum)
-        portdir = "%s/%s/%s" % (zfcpsysfs, self.devnum, self.wwpn)
-        unitadd = "%s/unit_add" % (portdir)
-        unitdir = "%s/%s" % (portdir, self.fcplun)
-        failed = "%s/failed" % (unitdir)
+    def _free_device(self):
+        """Remove the device from the I/O ignore list to make it visible to the system.
+
+        :raises: ValueError if the device cannot be removed from the I/O ignore list
+        """
 
-        if not os.path.exists(online):
+        if not os.path.exists(self._device_online_path):
             log.info("Freeing zFCP device %s", self.devnum)
             util.run_program(["zfcp_cio_free", "-d", self.devnum])
 
-        if not os.path.exists(online):
+        if not os.path.exists(self._device_online_path):
             raise ValueError(_("zFCP device %s not found, not even in device ignore list.") %
                              (self.devnum,))
 
+    def _set_zfcp_device_online(self):
+        """Set the zFCP device online.
+
+        :raises: ValueError if the device cannot be set online
+        """
+
         try:
-            f = open(online, "r")
-            devonline = f.readline().strip()
-            f.close()
+            with open(self._device_online_path) as f:
+                devonline = f.readline().strip()
             if devonline != "1":
-                logged_write_line_to_file(online, "1")
+                logged_write_line_to_file(self._device_online_path, "1")
         except OSError as e:
             raise ValueError(_("Could not set zFCP device %(devnum)s "
                                "online (%(e)s).")
                              % {'devnum': self.devnum, 'e': e})
 
+    def _set_zfcp_device_offline(self):
+        """Set the zFCP device offline.
+
+        :raises: ValueError if the device cannot be set offline
+        """
+
+        try:
+            logged_write_line_to_file(self._device_online_path, "0")
+        except OSError as e:
+            raise ValueError(_("Could not set zFCP device %(devnum)s "
+                               "offline (%(e)s).")
+                             % {'devnum': self.devnum, 'e': e})
+
+    def online_device(self):
+        """Initialize the device and make its storage block device(s) ready to use.
+
+        :returns: True if success
+        :raises: ValueError if the device cannot be initialized
+        """
+
+        self._free_device()
+        self._set_zfcp_device_online()
+        return True
+
+
+class ZFCPDevice(ZFCPDeviceBase):
```

Comment:
I intentionally kept this class name so that it doesn't cause any problems for possible projects using the previous code. This could be a problem for RHEL-8/9.

Reply:
> ZFCPDeviceFullPath could be done like this, what do you think? [216fb82](https://github.com/storaged-project/blivet/commit/216fb82c386449547f542b5ce43bca4bd99eccf6)

Yes, like this. Not sure how python works regarding the singleton `zfcp = zFCP()` assigning an object of either subclass to a (non?!-member) variable `d` and then invoking (virtual) member functions `d.online_device()` or  `d.offline_device()` and remembering the objects in a set `self.fcpdevs`. Maybe python infers `None` (or whatever the world's top-most unknown pseudo class would be) as type of `d` and as type of set members. In another more strictly-typed OO language, I might have used an abstract ZFCPDevice (instead of ZFCPDeviceBase), use it as type for `d`, and derive the two concrete ZFCPDevice... classes from it. IIRC, ZFCPDevice... is not used anywhere else in the code so it's really encapsulated here in the singleton and there should be no other depencies. The only used (by anaconda UI or pykickstart) interfaces of the singleton seem to be:

```
pyanaconda/modules/storage/kickstart.py:273:            zfcp.add_fcp(fcp.devnum, fcp.wwpn, fcp.fcplun)
pyanaconda/modules/storage/reset.py:78:            zfcp.startup()
pyanaconda/modules/storage/zfcp/discover.py:80:            zfcp.add_fcp(self._device_number, self._wwpn, self._lun)
pyanaconda/modules/storage/zfcp/zfcp.py:58:        zfcp.startup()
pyanaconda/modules/storage/zfcp/zfcp.py:78:        zfcp.write(conf.target.system_root)
```

This stuff here in the singleton is only for explicit online/offline and/or zfcp path management. In contrast, ZFCPDiskDevice exists for each SCSI disk device enumerated by libudev and backed by zfcp-attachment, so that's independent of ZFCPDevice... here. The stuff here triggers the appearance of the SCSI disk devices over there (handling `dracut_setup_args`).

> 
> But at this point I do not plan doing any other changes in the code due to other priorities.

Fair enough. Let's get in what we have.

Reply:
> Yes, like this. Not sure how python works regarding the singleton `zfcp = zFCP()` assigning an object of either subclass to a (non?!-member) variable `d` and then invoking (virtual) member functions `d.online_device()` or `d.offline_device()` and remembering the objects in a set `self.fcpdevs`. 

I tested the change before sending it and there was no issue when using the `zFCP()` singleton. But I could only test devices with auto LUN scan disabled (by `zfcp.allow_lun_scan=0`) or enabled, but not mixing both device types at the same time (if it's at all possible).

Anyway, I've updated https://github.com/storaged-project/blivet/pull/1061 with renaming the ZFCPDevice class.